Tests/ApolloCodegenTests/ApolloCodegenTests.swift

```diff
@@ -0,0 +1,193 @@
+import XCTest
+import ApolloCodegenTestSupport
+@testable import ApolloCodegenLib
+import Nimble
+
+class ApolloCodegenTests: XCTestCase {
+  private let schemaSDL: String = {
+    """
+    type Query {
+      books: [Book!]!
+      authors: [Author!]!
+    }
+
+    type Book {
+      title: String!
+      author: Author!
+    }
+
+    type Author {
+      name: String!
+      books: [Book!]!
+    }
+    """
+  }()
+
+  let directoryURL = CodegenTestHelper.outputFolderURL().appendingPathComponent("Codegen")
+
+  override func setUpWithError() throws {
+    try FileManager.default.apollo.createDirectoryIfNeeded(atPath: directoryURL.path)
+  }
+
+  override func tearDownWithError() throws {
+    try cleanTestOutput()
+  }
+
+  // MARK: Helpers
+
+  private func cleanTestOutput() throws {
+    try FileManager.default.apollo.deleteDirectory(atPath: directoryURL.path)
+  }
+
+  private func createSchema(at path: String) {
+    expect(
+      try FileManager.default.apollo.createFile(
+        atPath: path,
+        data: self.schemaSDL.data(using: .utf8)!
+      )
+    ).notTo(throwError())
+  }
+
+  private func createOperationFile(for data: Data, inDirectory directoryURL: URL) {
+    expect(
+      try FileManager.default.apollo.createFile(
+        atPath: directoryURL.appendingPathComponent("\(UUID().uuidString).graphql").path,
+        data: data
+      )
+    ).notTo(throwError())
+  }
+
+  // MARK: Tests
+
+  func test_build_givenInvalidConfiguration_shouldThrow() throws {
+    // given
+    let config = ApolloCodegenConfiguration(basePath: "not_a_path")
+
+    // then
+    expect(try ApolloCodegen.build(with: config)).to(throwError())
+  }
+
+  /// ```
+  /// SCHEMA:
+  /// type Query {
+  ///   books: [Book!]!
+  ///   authors: [Author!]!
+  /// }
+  ///
+  /// type Book {
+  ///   title: String!
+  ///   author: Author!
+  /// }
+  ///
+  /// type Author {
+  ///   name: String!
+  ///   books: [Book!]!
+  /// }
+  ///
+  /// OPERATION:
+  /// query getBooks {
+  ///   books {
+  ///     title
+  ///     name (this will cause the validation error)
+  ///   }
+  /// }
+  /// ```
```

Comment:
I like this comment from a theoretical perspective in terms of showing what's happening in this test, but I again worry about the comment and the actual test diverging at some point. I think given that the schema being used is right in this file throwing an explanation of the error into a `// when` comment might be a little more obvious.

Reply:
Yes, 100% agree. Not every test needs to (or should) be commented like this. I only did that because the actual setup for each test and the expected values were not easy to read and understand. 

In most cases, the tests should be self explanatory and clear just from reading them. Only when we feel like we can’t make that happen should we by commenting every test to this extent. 

I’m always fine with a quick little comment explaining why we are testing something or giving a little detail. But having the entire given and expected laid out in a comment like this is not a best practice I want to create. It’s fine when needed, but not a best practice. 

Like @designatednerd foretold, I’m already having issues with the comments and tests deviating in the tests I wrote like this, and I didn’t even know it. So this is not great. 

I’m thinking I could make those tests a little clearer to read without the comments if I created a result builder for it. But that’s a lot of work for a test utility.

Reply:
I went back and forth on leaving in this comment for exactly the reason of being brittle and divergence - it's not a best practice I agree.

> throwing an explanation of the error into a // when comment might be a little more obvious.

I like this suggestion, thanks.